wsmaster/che-core-api-devfile/src/main/java/org/eclipse/che/api/devfile/server/DevfileManager.java

```diff
@@ -0,0 +1,126 @@
+import javax.inject.Singleton;
+import org.eclipse.che.api.core.ConflictException;
+import org.eclipse.che.api.core.NotFoundException;
+import org.eclipse.che.api.core.ServerException;
+import org.eclipse.che.api.core.ValidationException;
+import org.eclipse.che.api.devfile.model.Devfile;
+import org.eclipse.che.api.devfile.server.validator.DevfileIntegrityValidator;
+import org.eclipse.che.api.devfile.server.validator.DevfileSchemaValidator;
+import org.eclipse.che.api.workspace.server.WorkspaceManager;
+import org.eclipse.che.api.workspace.server.model.impl.WorkspaceConfigImpl;
+import org.eclipse.che.api.workspace.server.model.impl.WorkspaceImpl;
+import org.eclipse.che.commons.env.EnvironmentContext;
+
+@Singleton
+public class DevfileManager {
+
+  private final ObjectMapper objectMapper;
+  private DevfileSchemaValidator schemaValidator;
+  private DevfileIntegrityValidator integrityValidator;
+  private DevfileConverter devfileConverter;
+  private WorkspaceManager workspaceManager;
+
+  @Inject
+  public DevfileManager(
+      DevfileSchemaValidator schemaValidator,
+      DevfileIntegrityValidator integrityValidator,
+      DevfileConverter devfileConverter,
+      WorkspaceManager workspaceManager) {
+    this.schemaValidator = schemaValidator;
+    this.integrityValidator = integrityValidator;
+    this.devfileConverter = devfileConverter;
+    this.workspaceManager = workspaceManager;
+    this.objectMapper = new ObjectMapper(new YAMLFactory());
+  }
+
+  /**
+   * Creates {@link WorkspaceConfigImpl} from given devfile content. Performs schema and integrity
+   * validation of input data before conversion.
+   *
+   * @param devfileContent raw content of devfile
+   * @param verbose when true, method returns more explained validation error messages if any
+   * @return WorkspaceConfig created from the devfile
+   * @throws DevfileFormatException when any of schema or integrity validations fail
+   * @throws JsonProcessingException when parsing error occurs
+   */
+  public WorkspaceConfigImpl convert(String devfileContent, boolean verbose)
+      throws DevfileFormatException, JsonProcessingException {
+    JsonNode parsed = schemaValidator.validateBySchema(devfileContent, verbose);
+    Devfile devFile = objectMapper.treeToValue(parsed, Devfile.class);
+    integrityValidator.validateDevfile(devFile);
+    return devfileConverter.devFileToWorkspaceConfig(devFile);
+  }
+
+  /**
+   * Creates workspace from given {@link WorkspaceConfigImpl} with available name search
+   *
+   * @param workspaceConfig initial workspace configuration
+   * @return created workspace instance
+   * @throws ValidationException when incoming configuration or attributes are not valid
+   * @throws ConflictException when any conflict occurs
+   * @throws NotFoundException when user account is not found
+   * @throws ServerException when other error occurs
+   */
+  public WorkspaceImpl createWorkspace(WorkspaceConfigImpl workspaceConfig)
```

Comment:
For me it looks strange to see such public method in `DevfileManager`. As far as I see it is used in Service in the following way:
```
  ....
 WorkspaceConfig config = manager.convert(devfileContent, verbose);
 manager.createWorkspace(config);
```
I see it more logical to have public method `WorkspaceImpl createWorkspace(String devfileContent)` in `DevfileManager`

Reply:
Too much responsibility for single method and not flexible in usage IMO.
I would rather propose another schema:
1) Devfile content to devfile object - with schema validation but without intergrity validation;
2) Create WS from devfile - integrity validation, then convert to WsConfig and create WS. With this way, it will be easy to add later additional optional params for stuff like K8S list etc to be merged together into config. WDYT ?

Reply:
yes, that's it.